### PR TITLE
Adds Modal

### PR DESCRIPTION
**Pull request by jbrundage**
@GordonSmith please review

Signed-off-by: Jay Brundage <jaman.brundage@lexisnexis.com>

### CODE REVIEW

File: packages/layout/src/Modal.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+    getRelativeTarget() {
+        let relativeTarget;
+        if (this.relativeTargetId()) {
+            relativeTarget = document.getElementById(this.relativeTargetId());
+            if (relativeTarget) {
+                return relativeTarget;
+            }
+        }
+        if (!relativeTarget) {
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: if not needed.

File: packages/layout/src/Modal.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                .style("max-width", this.maxWidth())
+                ;
+        }
+    }
+
+    setFadePosition(rect) {
+        this._fade.node().style.top = rect.top + "px";
+        this._fade.node().style.left = rect.left + "px";
+        this._fade.node().style.width = rect.width + "px";
+        this._fade.node().style.height = rect.height + "px";
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: For consistency change to:
```
this
  ._fade.style("top", `${rect.top}px`)
  ...
  ;
```

File: packages/layout/src/Modal.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        const contentRect = _node.getBoundingClientRect();
+        _node.style.position = "fixed";
+        _node.style.top = (rect.top + (rect.height / 2) - (contentRect.height / 2)) + "px";
+        _node.style.left = (rect.left + (rect.width / 2) - (contentRect.width / 2)) + "px";
+        _node.style.width = contentRect.width + "px";
+        _node.style.height = contentRect.height + "px";
+        if (this.minHeight() || this.minWidth()) {
+            _node.style["min-height"] = this.minHeight();
+            _node.style["min-width"] = this.minWidth();
+        }
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: For consistency change to:
```
this
  ._modal.style("XXX", YYY)
  ...
  ;
```

File: packages/layout/src/Modal.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    enter(domNode, element) {
+        super.enter(domNode, element);
+        const context = this;
+        this._fade = element.append("div")
+            .classed("layout_Modal-fade", true)
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: It should not need the "layout_Modal" prefix as that is already assigned to a parent div, so out selectors can still be specific: `.layout_Modal > div.fade` type of thing?

File: packages/layout/src/Modal.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            ;
+        this._modalHeader = this._modal.append("div")
+            .classed("layout_Modal-header", true)
+            .style("color", this.titleFontColor())
+            .style("font-size", this.titleFontSize() + "px")
+            .style("height", header_h + "px")
+            ;
+        this._modalBody = this._modal.append("div")
+            .classed("layout_Modal-body", true)
+            .style("height", `calc( 100% - ${header_h}px )`)
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: I just checked MDN and don't see any reason not to use `calc` - which will potentially help simplify layout code in the future!

File: packages/layout/src/Modal.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                    n.resize().render();
+                })
+                ;
+        } else {
+            this._widget
+                .target(null)
+                .render()
+                ;
+        }
+        this.element().style("display", this.show() ? null : "none");
```
Comment (GordonSmith) [inline packages/layout/src/Modal.ts]: You can just use `element` instead of `this.element()`

File: tests/test-layout/src/layout.spec.ts [hunk 7]
```diff
@@ -196,6 +196,16 @@ describe("@hpcc-js/layout", function () {
...
                                 break;
+                            case Modal:
+                                render(
+                                    new Modal()
+                                        .widget(
+                                        new Pie()
+                                            .columns(data.TwoD.subjects.columns)
+                                            .data(data.TwoD.subjects.data)
+                                        )
+                                );
```
Comment (GordonSmith) [inline tests/test-layout/src/layout.spec.ts]: Indentation is off.

File: packages/layout/src/Modal.css [hunk 8]
```diff
@@ -0,0 +1,48 @@
+.layout_Modal * {
+    box-sizing: border-box;
+}
```
[the remark this answers is not in the document]
Comment (jbrundage) [reply packages/layout/src/Modal.css]: Potentially... I'll remove it, but the safest approach would be avoiding all use of border and padding.

IMO all html should use `box-sizing: border-box`

File: packages/layout/src/Modal.ts [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        super();
+        this._tag = "div";
+    }
+
+    closeModal() {
+        if (typeof this._close === "function") {
+            this._close();
+        }
+        this.show(false).render();
+    }
```
[the remark this answers is not in the document]
Comment (jbrundage) [reply packages/layout/src/Modal.ts]: Wow... can "on" simply hook onto any class method like that?

[the remark this answers is not in the document]
Comment (GordonSmith) [reply packages/layout/src/Modal.ts]: Yes - this is all it does:  https://github.com/hpcc-systems/Visualization/blob/master/src/common/Widget.js#L89-L103

File: packages/layout/src/Modal.ts (hunk 4 above)
Comment (jbrundage) [reply packages/layout/src/Modal.ts]: `.fade` is too common of a name. Could cause issues.

**Comment by GordonSmith** [comment]
Closing in favor of:  https://github.com/hpcc-systems/Visualization/pull/2356

**Pull request closed without merging**